### PR TITLE
Accessibility: Add disposable focus Higher Order Component

### LINKED ISSUE #895: Focus management: a practical example of focus loss in the Sidebar
There are already some issues (see for example #297 and #515) about focus management (that's of fundamental importance for accessibility) and also some discussions have already happened.

In this issue I'd like to clarify a bit the effect of a focus loss and what that does mean for users who don't use a pointing device. I've made a short video (22 seconds!) using the current Gutenberg plugin as of May 23rd: https://cloudup.com/c-ls4zlBvk0

A few notes:
- I'm using a custom plugin I use for debugging a11y, that adds borders and outlines to all the `focusable` elements: this breaks a bit the layout, please ignore the breakage
- once on the "Post Settings" button, there's no way to tab into the sidebar, so I'm using the mouse to move focus there
- then I tab to the "X" close button and press Enter or Spacebar to activate it
- at that point, the element that was focused gets removed from the DOM
- tab again, and the browser moves focus back to the document root: this happens even if Chrome has implemented a sort of "internal pointer" to try to keep focus in place since version 50 (Firefox always had a similar feature, Safari form version 10)

This focus loss actually forces to start navigating content from scratch, which is one of the worst thing can happen for users who don't use a pointing device.

To avoid this focus losses there's really the need of a "focus management" tool able to get references to DOM elements, even if they pertain to different components. This tool should give the ability to move focus back in a logical place when necessary, to evaluate on a case by case basis.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #895 

This PR adds a generic `disposableFocus` higher order component.
A component wrapped in the "disposableFocus" hoc will have the following behaviour:
 - Saves the active element when the disposable element is mounted
 - Restores the active element when the disposable element is unmounted (unless the focus has moved outside the disposable component)

All disposable elements should be wrapped within this HoC: Thinking of inserter, sidebar...

### CODE REVIEW

**Comment by jasmussen** [comment]
Thanks for working on this. I'm not entirely convinced how to test this exactly, so would appreciate @afercia's eyes as well. The inserter seems to be working well (outside of that focus rectangle which is my fault, and for which I'm opening a separate ticket for soon).

Codewise, I like all the red I see this clearing up!

**Comment by afercia** [comment]
Looks very promising to me, as it's a simple way to get an UI control to return focus to later. Tested on both the Inserter and Sidebar, and focus is correctly moved back to the related buttons. About the Sidebar for example, it solves the focus loss pointed out in #895 

I'd like to see this as the first tool of a more complete "focus management" set of tools.

Just one thing, I'd consider to don't use a `<div>` element for the wrapper, since it could happen to use it in places where a `<div>` is not allowed and would cause invalid HTML. After all this wrapper is not meant for layout purposes and could be, for example, a span.

Not sure about the naming 🙂  maybe `disposableFocus` is not so clear, but I wouldn't know what other naming to use. Minor point anyways.

**Comment by youknowriad** [comment]
> Just one thing, I'd consider to don't use a <div> element for the wrapper

Unfortunately, it's not possible to avoid a wrapper here. The wrapper tagName could be made customizable though (if we need to). We need a wrapper to get a reference to the dom node.

**Comment by youknowriad** [comment]
Actually, I was able to drop the wrapper in 74f0dc8

**Comment by afercia** [comment]
Sorry, just noticed a drawback: when clicking using the mouse, `document.activeElement` returns the `body` element in Firefox and Safari. Returns the button in Chrome. Haven't tested in IE.

The previous method used for the inserter worked also when clicking with the mouse, because it was using the click event current target.

It would be great to find a way to make this work both with mouse and keyboard!

**Comment by youknowriad** [comment]
@afercia Good catch! I noticed that this happening because when we click on the "Insert" Button for example, the focus is not being moved to the button in Firefox and Safari while it's done in Chrome. I can't find yet why the focus is not moved to buttons when clicking on them.

**Comment by youknowriad** [comment]
I found this link showing these differences. thinking how can we make this work the Chrome way everywhere https://gist.github.com/cvrebert/68659d0333a578d75372

File: components/disposable-focus/index.js
Comment (aduth) [inline components/disposable-focus/index.js]: `findDOMNode` is [documented](https://facebook.github.io/react/docs/react-dom.html#finddomnode) as an escape hatch, though in this case I think a reasonable one. Concerns me to see [some conversation](https://twitter.com/dan_abramov/status/752936646602031104) in which React maintainers seem obviously inclined to put it on a path toward deprecation.

I noted the earlier point about wrapper components. I think we should keep this as-is, but leave open as options for future consideration:

- Wrap with `span` to reduce layout implications (inline, not block)
- Expect the mounted component to define a node getter (similar to [`react-click-outside`'s `handleClickOutside`](https://github.com/kentor/react-click-outside/blob/4934ef15531f3a0267e243aac39ed96dc10dfe45/index.js#L28))
- Pass callback as prop to the wrapped component, expected to be called on mount with reference to the DOM node

**Comment by youknowriad** [comment]
@afercia Do you agree that buttons should receive the focus when we click on them all the time (like chrome does)? In that case, we can add a `onClick` handle to focus the buttons in a generic way in the `Button` component to mimic the Chrome's behaviour?

Thoughts on this @aduth?

File: components/disposable-focus/index.js
Comment (aduth) [inline components/disposable-focus/index.js]: Re: Folder placement. This is technically not a component, but rather a function which returns a modified component. Dunno if or how we'd want to distinguish this.

File: editor/inserter/index.js
Comment (aduth) [inline editor/inserter/index.js]: I think some of the inline documentation could be adapted for use in the new higher-order variant.

File: components/disposable-focus/index.js
Comment (youknowriad) [inline components/disposable-focus/index.js]: I was thinking of providing a config object as a parameter to the HoC as an alternative.

`disposableFocus({ wrapper: 'span' })(Component)`

**Comment by afercia** [comment]
> Do you agree that buttons should receive the focus when we click on them all the time (like chrome does)?

Just tested Firefox on Windows and `document.activeElement` correctly returns the buttons also when clicking with the mouse. 😬  That confirms what reported in the link you posted, data actually comes from MDN, see: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus

Ideally, I think it would make sense to try to pair this behaviour across platforms/browsers.

**Comment by afercia** [comment]
Tested also in:
Edge: works
IE 11: can't test, still getting `Object doesn't support property or method 'from'`, see #664

**Comment by youknowriad** [comment]
I'm thinking we should merge this. And try (or not) to fix the browser issue separately. The fact that this issue breaks the `disposableFocus` is just a consequence and the HoC is not doing anything wrong.

**Comment by aduth** [comment]
@youknowriad What do we want to do about these review points?

- https://github.com/WordPress/gutenberg/pull/901/files#r118479603
- https://github.com/WordPress/gutenberg/pull/901/files#r118479706

**Comment by youknowriad** [comment]
@aduth Oh sorry, I forgot about the documentation comment.

About the `findDomNode` I'm thinking we should leave it for now, since it's not deprecated or anything yet but keeping the possibility to add the wrapper back with a config if this changes. What do you think?

**Comment by youknowriad** [comment]
@aduth 😅 I need a good long weekend :P 

For the placement, I don't know, I'm open to suggestions. I put it in "components" to make it reusable across other WordPress modules (outside Editor/Blocks), maybe we can create a subfolder `components/hoc`

**Comment by aduth** [comment]
Yeah, I don't really have a great answer. It's not a true component, but I don't necessarily want another top-level directory for these functions. I think they're _related_ to components and in that way could belong in the components directory, but I find the `hoc` acronym to be a fair bit cryptic. It would be nice to have something short for plugin consumers using the global `wp.components.hoc`. I'm wondering if `components/higher-order/` (`wp.components.higherOrder`) would be an "okay-ish" compromise.

**Comment by youknowriad** [comment]
Let's move it to the `higher-order` subfolder for now and we'll see how it feels with other HoCs :)

Also, thinking we should add a`index.js` to the `components` and the `higher-order` folder to allow `import { Button, IconButton } from 'components' and `import { disposableFocus } from 'components/higher-order'`. But let's leave this to a separate PR

**Comment by aduth** [comment]
>Also, thinking we should add a`index.js` to the `components` and the `higher-order` folder to allow `import { Button, IconButton } from 'components' and `import { disposableFocus } from 'components/higher-order'`. But let's leave this to a separate PR

Yep, this is needed too for exposing components on the global. I was wondering if there might be a way to automate it, so we can avoid [this kind of scenario](https://github.com/Automattic/wp-calypso/blob/master/client/state/selectors/index.js).

**Comment by aduth** [comment]
>Not sure about the naming 🙂  maybe `disposableFocus` is not so clear, but I wouldn't know what other naming to use. Minor point anyways.

Maybe something about focus being remembered or returned? `focusMemory`, `withFocusReturn` ?

**Review by aduth — APPROVED** [review]

File: components/higher-order/with-focus-return/index.js
Comment (mtias) [inline components/higher-order/with-focus-return/index.js]: What about `utilities` instead of higher-order? Alternatively, we can just keep them in the root. cc @aduth @youknowriad

**Comment by nylen** [comment]
Why is this component named `withFocusReturn` instead of `WithFocusReturn`?  I would think that everything under `components` should be named in upper-case.

**Comment by youknowriad** [comment]
@nylen to make the distinction that it's not a component, it's a component factory. It can't be used like this `<WithReturnFocus />`

File: components/higher-order/with-focus-return/index.js
Comment (youknowriad) [inline components/higher-order/with-focus-return/index.js]: I don't have a preference here, happy to update to what we'll settle on.

Comment (aduth) [inline components/higher-order/with-focus-return/index.js]: I find "utilities" to be a meaningless word.

**Pull request merged**
